projects/onto-search/src/lib/onto-search-facet/datepicker-facet/datepicker-facet.component.ts

```diff
@@ -0,0 +1,182 @@
+import {Component, EventEmitter, Input, Output, Renderer2, TemplateRef} from '@angular/core';
+import {OntoSearchFacetComponent} from '../onto-search-facet.component';
+import {FormControl, FormGroup} from '@angular/forms';
+import {SearchDateFacetGroupModel} from './models/search-date-facet-group-model';
+import moment, {Moment} from 'moment';
+import {DateRange} from '@angular/material/datepicker';
+import {SearchDateFacetModel} from './models/search-date-facet-model';
+import {SearchDateFacetRange} from './models/search-date-facet-range';
+
+@Component({
+  selector: 'onto-datepicker-facet',
+  templateUrl: './datepicker-facet.component.html',
+  styleUrls: ['./datepicker-facet.component.scss']
+})
+export class DatepickerFacetComponent extends OntoSearchFacetComponent {
+  @Input()
+  public data: SearchDateFacetGroupModel;
+
+  @Input()
+  public facetTemplate: TemplateRef<any>;
+
+  @Input()
+  public facetTitleTemplate: TemplateRef<any>;
+
+  @Output()
+  public onSelectionChange: EventEmitter<SearchDateFacetRange> = new EventEmitter<SearchDateFacetRange>();
+
+  dateRangeGroup: FormGroup = new FormGroup({
+    start: new FormControl(),
+    end: new FormControl()
+  });
+
+  datePickerPlaceholder: string;
+
+  private daysMap: Map<string, number> = new Map<string, number>();
+  private monthsMap: Map<string, number> = new Map<string, number>();
+  private yearsMap: Map<string, number> = new Map<string, number>();
+
+  private buttonUnsubscribeFns: (() => void)[] = [];
+
+  constructor(private renderer: Renderer2) {
+    super();
+  }
+
+  ngOnInit(): void {
+    super.ngOnInit();
+    this.mapFacetDates(this.data.facetGroupData);
+
+    this.datePickerPlaceholder = this.data.placeholder;
+    if (this.data.selectedRange) {
+      this.setSelectedRange(this.data.selectedRange);
+    }
+  }
+
+  calendarOpened(event): void {
```

Comment:
What kind of sorcery is this method doing?

Reply:
I really prefer any setTimeout/setInterval usage to be properly documented. In most cases which i"ve seen excluding some optimization attempts, these are used in some dirty hacks. That's why it's good to have some explanations.

Reply:
I will add a comment about the setTimeout. It is used to that the logic is applied after all rendering is done.
What the method does is get all buttons from the calendar and add listeners on click and updates the styles of those that have result count in the maps (dayMap, monthMap, yeraMap).
The listeners are used to detect view change of the calendar (from month to year or multiyear).
When the calendar is closed, all listeners are unsubscribed.